File: 47. Permutations II/_47.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = list()
        self.dfs(nums, res, [])
        return res

    def dfs(self, nums, res, tmp):
        if len(nums) == 0:
            res.append(tmp)
            return
        for i in range(len(nums)):
            if i == 0:
                pre = nums[0]
                self.dfs(nums[1:], res, tmp+[nums[0]])

            else:
                if nums[i] == pre:
                    continue

                else:
                    pre = nums[i]
                    self.dfs(nums[:i]+nums[i+1:], res, tmp+[nums[i]])

File: 47. Permutations II/47.go
package leetcode

import(
	"sort"
)

func permuteUnique(nums []int) [][]int {
	sort.Ints(nums)
	res := make([][]int, 0)
	dfs(&res, nums, []int{})
	return res
}

func dfs(res *[][]int, nums []int, tmp []int){
	if len(nums) == 0{
		*res = append(*res, tmp)
	}

	var pre int
	for i,v := range nums{
		newtmp := make([]int, 0, len(tmp)+1)
		newnums := make([]int, 0, len(nums)-1)
		if i == 0{
			pre = nums[0]			
			newtmp = append(newtmp, tmp...)
			newtmp = append(newtmp, v)
			newnums = append(newnums, nums[1:]...)
			dfs(res, newnums, newtmp)
		}else{
			if v == pre{
				continue
			}else{
				pre = v
				newtmp = append(newtmp, tmp...)
				newtmp = append(newtmp, v)
				newnums = append(newnums, nums[:i]...)
				newnums = append(newnums, nums[i+1:]...)
				dfs(res, newnums, newtmp)
			}
		}
	}
}